# Data Mining – Assignment 1

## Mining Big Datasets

---

> Maria Zafeiropoulou & Marianna Konstantopoulou <br />
> P2822113 & P2822122 <br />
> MSc Business Analytics Part Time <br />

## Questions

### 1) Import and pre-process the dataset with customers

In [2]:
import pandas as pd
import numpy as np

* Importing the csv file:

In [3]:
bank = pd.read_csv('bank.csv',sep=';')
bank

Age           Job   Marital  Education Default  Balance Housing Loan  \
0      33.0  entrepreneur   married  secondary      no        2     yes  yes   
1      35.0    management   married   tertiary      no      231     yes   no   
2       NaN    management    single   tertiary      no      447     yes  yes   
3      42.0  entrepreneur  divorced   tertiary     yes        2     yes   no   
4      58.0       retired   married    primary      no      121     yes   no   
...     ...           ...       ...        ...     ...      ...     ...  ...   
43186  51.0    technician   married   tertiary      no      825      no   no   
43187  71.0       retired  divorced    primary      no     1729      no   no   
43188  72.0       retired   married  secondary      no     5715      no   no   
43189  57.0   blue-collar   married  secondary      no      668      no   no   
43190  37.0  entrepreneur   married  secondary      no     2971      no   no   

          Rating                   Products  
0           poor               1,3,16,17,19  
1           good                     4,8,16  
2           fair                       7,16  
3           fair       1,3,8,10,11,12,18,19  
4           good           4,5,6,7,11,18,19  
...          ...                        ...  
43186  very_good                      12,15  
43187  very_good                    4,17,19  
43188  very_good  5,7,8,9,10,11,13,16,17,18  
43189  very_good                      2,3,7  
43190  very_good           3,6,7,9,12,15,18  

[43191 rows x 10 columns]

In [5]:
bank.head(10)

Age           Job   Marital  Education Default  Balance Housing Loan  \
0  33.0  entrepreneur   married  secondary      no        2     yes  yes   
1  35.0    management   married   tertiary      no      231     yes   no   
2   NaN    management    single   tertiary      no      447     yes  yes   
3  42.0  entrepreneur  divorced   tertiary     yes        2     yes   no   
4  58.0       retired   married    primary      no      121     yes   no   
5  43.0    technician    single  secondary      no      593     yes   no   
6  41.0        admin.  divorced  secondary      no      270     yes   no   
7  29.0        admin.    single  secondary      no      390     yes   no   
8  53.0    technician   married  secondary      no        6     yes   no   
9   NaN      services   married  secondary      no      162     yes   no   

  Rating                  Products  
0   poor              1,3,16,17,19  
1   good                    4,8,16  
2   fair                      7,16  
3   fair      1,3,8,10,11,12,18,19  
4   good          4,5,6,7,11,18,19  
5   good    1,5,7,8,11,16,17,19,20  
6   good  2,3,4,5,9,10,11,12,14,17  
7   good                        17  
8   good    8,13,14,16,17,18,19,20  
9   good                         3

* Using `describe()` we will check if our numeric values have correct ranges.

In [6]:
round(bank.describe(),2)

Age    Balance
count  42191.00   43191.00
mean      40.79    1354.04
std       10.54    3042.17
min       18.00   -8019.00
25%       33.00      71.00
50%       39.00     442.00
75%       48.00    1412.00
max       95.00  102127.00

* Next step would be to check if we have any missing values.

In [7]:
pd.isnull(bank).sum()

Age          1000
Job             0
Marital         0
Education       0
Default         0
Balance         0
Housing         0
Loan            0
Rating          0
Products        0
dtype: int64

* We replace the missing values at `Age` with the average value (rounded to the nearest integer)

In [8]:
bank['Age'].fillna(value=round(bank['Age'].mean()), inplace=True)

* The mean rounded was 41. Let's check one more time for missing values.

In [9]:
pd.isnull(bank).sum()

Age          0
Job          0
Marital      0
Education    0
Default      0
Balance      0
Housing      0
Loan         0
Rating       0
Products     0
dtype: int64

Replacing the `Rating` values with numbers from 1-5 since it's an ordinal variable.

In [10]:
bank=bank.replace(to_replace="poor",value="1")
bank=bank.replace(to_replace="fair",value="2")
bank=bank.replace(to_replace="good",value="3")
bank=bank.replace(to_replace="very_good",value="4")
bank=bank.replace(to_replace="excelent",value="5")

In [11]:
bank

Age           Job   Marital  Education Default  Balance Housing Loan  \
0      33.0  entrepreneur   married  secondary      no        2     yes  yes   
1      35.0    management   married   tertiary      no      231     yes   no   
2      41.0    management    single   tertiary      no      447     yes  yes   
3      42.0  entrepreneur  divorced   tertiary     yes        2     yes   no   
4      58.0       retired   married    primary      no      121     yes   no   
...     ...           ...       ...        ...     ...      ...     ...  ...   
43186  51.0    technician   married   tertiary      no      825      no   no   
43187  71.0       retired  divorced    primary      no     1729      no   no   
43188  72.0       retired   married  secondary      no     5715      no   no   
43189  57.0   blue-collar   married  secondary      no      668      no   no   
43190  37.0  entrepreneur   married  secondary      no     2971      no   no   

      Rating                   Products  
0          1               1,3,16,17,19  
1          3                     4,8,16  
2          2                       7,16  
3          2       1,3,8,10,11,12,18,19  
4          3           4,5,6,7,11,18,19  
...      ...                        ...  
43186      4                      12,15  
43187      4                    4,17,19  
43188      4  5,7,8,9,10,11,13,16,17,18  
43189      4                      2,3,7  
43190      4           3,6,7,9,12,15,18  

[43191 rows x 10 columns]

### 2) Compute data (dis-)similarity

To assess the similarity between the customers you could form the dissimilarity matrix for all given attributes. As described in lecture “Measuring Data Similarity”, for every given attribute you first distinguish its type (categorical, ordinal, numerical or set) and then compute the dissimilarity of its values accordingly. For set similarity use the Jaccard similarity between sets. Then, you can calculate the average of the computed dissimilarities to derive the dissimilarity over all attributes. Depending on the machine used to implement this assignment you should decide whether it is feasible to compute the dissimilarity matrices, or, have the computations performed on-the-fly for a pair of customers.

* Firstly, we need to create functions in order to compute the dissimilarities.

* The first function `dis_categorical` computes the dissimilarity for categorical variables:

In [12]:
def dis_categorical(a,b,col):
    if bank[col].iloc[a]==bank[col].iloc[b]: #If a=b then d(a,b)=0 else if a!=b then d(a,b)=1
        return 0
    else:
        return 1
    
dis_categorical(1,2,'Job') #Testing if our function works for 2 costumers

0

* The second function `dis_ordinal` computes the dissimilarity for ordinal variables:

In [13]:
def dis_ordinal(a,b):
    return abs((int(bank['Rating'].iloc[a]))-(int(bank['Rating'].iloc[b])))/(int(max(bank['Rating']))-int(min(bank['Rating'])))
#This function computes for a,b ordinal then d(a,b) = |rank(a) - rank(b)| / maxrank - minrank
dis_ordinal(1,2)

0.25

* The third function `dis_numeric` computes the dissimilarity for numeric variables:

In [14]:
def dis_numeric(a,b,col):
    return abs((bank[col].iloc[a])-(bank[col].iloc[b]))/(int(max(bank[col]))-int(min(bank[col])))
#This function computes for a,b numeric then d(a,b) = |a-b| / maxvalue - minvalue
dis_numeric(1,2,'Age')

0.07792207792207792

* Finally, for sets we need to create function `dis_jaccard` that uses the Jaccard Index:

In [15]:
def dis_jaccard(a, b):
    #Find intersection of two sets
    nominator = set(bank['Products'].iloc[a].split(',')).intersection(set(bank['Products'].iloc[b].split(',')))
    #Find union of two sets
    denominator = set(bank['Products'].iloc[a].split(',')).union(set(bank['Products'].iloc[b].split(',')))
    #Take the 1-ratio of sizes
    dissimilarity = 1-(len(nominator)/len(denominator))
    return dissimilarity
#Jaccard(S1,S2) =|S1 ∩ S2|/|S1 ∪ S2|
dis_jaccard(1,2)

0.75

* Now, we will have to create blank matrices for all dissimilarity matrices (one for each column) and then we need to fill them with the correct figures. For performance reasons we will have the computations performed on-the-fly for a pair of customers:

In [16]:
#Creating the blank matrices for each variable
age_dis_matrix=np.zeros(shape=(2,2))
job_dis_matrix=np.zeros(shape=(2,2))
marital_dis_matrix=np.zeros(shape=(2,2))
education_dis_matrix=np.zeros(shape=(2,2))
default_dis_matrix=np.zeros(shape=(2,2))
balance_dis_matrix=np.zeros(shape=(2,2))
housing_dis_matrix=np.zeros(shape=(2,2))
loan_dis_matrix=np.zeros(shape=(2,2))
rating_dis_matrix=np.zeros(shape=(2,2))
products_dis_matrix=np.zeros(shape=(2,2))

#Filling the matrices for two customers using the functions we created earlier:
for i in range(2):
    for j in range(2):
        age_dis_matrix[i,j] = dis_numeric(i,j,'Age')
        job_dis_matrix[i,j] = dis_categorical(i,j,'Job')
        marital_dis_matrix[i,j] = dis_categorical(i,j,'Marital')
        education_dis_matrix[i,j] = dis_categorical(i,j,'Education')
        default_dis_matrix[i,j] = dis_categorical(i,j,'Default')
        balance_dis_matrix[i,j] = dis_numeric(i,j,'Balance')
        housing_dis_matrix[i,j] = dis_categorical(i,j,'Housing')
        loan_dis_matrix[i,j] = dis_categorical(i,j,'Loan')
        rating_dis_matrix[i,j] = dis_ordinal(i,j)
        products_dis_matrix[i,j] = dis_jaccard(i,j)
        
print("The Age dissimilarity matrix is:\n", age_dis_matrix)
print("The Job dissimilarity matrix is:\n",job_dis_matrix)
print("The Marital dissimilarity matrix is:\n", marital_dis_matrix)
print("The Education dissimilarity matrix is:\n", education_dis_matrix)
print("The Default dissimilarity matrix is:\n", default_dis_matrix)
print("The Balance dissimilarity matrix is:\n", balance_dis_matrix)
print("The Housing dissimilarity matrix is:\n", housing_dis_matrix)
print("The Loan dissimilarity matrix is:\n", loan_dis_matrix)
print("The Customer Rating dissimilarity matrix is:\n", rating_dis_matrix)
print("The Products dissimilarity matrix is:\n", products_dis_matrix)

#Combining scores by taking the average of computed dissimilarities
print("Combining scores:\n", (1/10)* (age_dis_matrix + job_dis_matrix + marital_dis_matrix + education_dis_matrix
                                        + default_dis_matrix + balance_dis_matrix + housing_dis_matrix +
                                        loan_dis_matrix + rating_dis_matrix + products_dis_matrix))

The Age dissimilarity matrix is:
 [[0.         0.02597403]
 [0.02597403 0.        ]]
The Job dissimilarity matrix is:
 [[0. 1.]
 [1. 0.]]
The Marital dissimilarity matrix is:
 [[0. 0.]
 [0. 0.]]
The Education dissimilarity matrix is:
 [[0. 1.]
 [1. 0.]]
The Default dissimilarity matrix is:
 [[0. 0.]
 [0. 0.]]
The Balance dissimilarity matrix is:
 [[0.         0.00207906]
 [0.00207906 0.        ]]
The Housing dissimilarity matrix is:
 [[0. 0.]
 [0. 0.]]
The Loan dissimilarity matrix is:
 [[0. 1.]
 [1. 0.]]
The Customer Rating dissimilarity matrix is:
 [[0.  0.5]
 [0.5 0. ]]
The Products dissimilarity matrix is:
 [[0.         0.85714286]
 [0.85714286 0.        ]]
Combining scores:
 [[0.         0.43851959]
 [0.43851959 0.        ]]


### 3) Nearest Neighbor (NN) search


Using the implementation of the previous step, you will calculate the 10-NN (most
similar)customers

* We will create a function which takes as an input the customer id and returns the 10 most similar customers and their similarity score. Using the same logic as we did in the previous step we create blank matrices for all dissimilarity matrices (one for each column) and then fill them with the correct figures using the fuctions creates in the previous steps as well. We combine scores by taking the average of computed dissimilarities minus 1 to result in the similarity score and finally we create a data frame with 10NN customers along with the corresponding similarity score.

In [17]:
def nn_search(cust_id):

 x = len(bank)

 #Creating the blank matrices for each variable
 age_dis_matrix=np.zeros(shape=(1,x))
 job_dis_matrix=np.zeros(shape=(1,x))
 marital_dis_matrix=np.zeros(shape=(1,x))
 education_dis_matrix=np.zeros(shape=(1,x))
 default_dis_matrix=np.zeros(shape=(1,x))
 balance_dis_matrix=np.zeros(shape=(1,x))
 housing_dis_matrix=np.zeros(shape=(1,x))
 loan_dis_matrix=np.zeros(shape=(1,x))
 rating_dis_matrix=np.zeros(shape=(1,x))
 products_dis_matrix=np.zeros(shape=(1,x))

 i = cust_id

#Filling the matrices for two customers using the functions we created earlier:
 for j in range(x): 
        age_dis_matrix[0,j] = dis_numeric(i-1,j,'Age')
        job_dis_matrix[0,j] = dis_categorical(i-1,j,'Job')
        marital_dis_matrix[0,j] = dis_categorical(i-1,j,'Marital')
        education_dis_matrix[0,j] = dis_categorical(i-1,j,'Education')
        default_dis_matrix[0,j] = dis_categorical(i-1,j,'Default')
        balance_dis_matrix[0,j] = dis_numeric(i-1,j,'Balance')
        housing_dis_matrix[0,j] = dis_categorical(i-1,j,'Housing')
        loan_dis_matrix[0,j] = dis_categorical(i-1,j,'Loan')
        rating_dis_matrix[0,j] = dis_ordinal(i-1,j)
        products_dis_matrix[0,j] = dis_jaccard(i-1,j)
    
#Calculating the similarity score
 comb = 1 - ((1/10)* (age_dis_matrix + job_dis_matrix + marital_dis_matrix + education_dis_matrix
                                        + default_dis_matrix + balance_dis_matrix + housing_dis_matrix +
                                         loan_dis_matrix + rating_dis_matrix + products_dis_matrix))

 
# Creating a dataframe with the similarity score
 comb = pd.DataFrame(np.transpose(comb))

# Adding the customer id
 comb['Customer ID'] = np.arange(comb.shape[0])+1
 comb.rename(columns = {0:'Similarity Score'}, inplace = True)

# Excluding the id of the target customer
 comb = comb.drop(comb[comb['Customer ID'] == i].index)

#Sorting the resulted data frame using the similarity score
 comb = comb.sort_values(by=['Similarity Score'], ascending = False)
# Keeping only the 10 first records 
 comb = comb.head(10)
#Change the column order of the dataframe 
 comb = comb.reindex(columns=['Customer ID','Similarity Score'])

#Print and return the wanted result
 print ('10 NN for Customer', i)
 return (comb)

* Using the nn_search function we will calculate the 10-NN customers for the following customer ids :<br>

<ol> 
    <li>1200<br> </li>
    <li>3650<br> </li>
    <li>10400<br></li>
    <li>14930<br></li>
    <li>22330<br></li>
    <li>25671<br></li>
    <li>29311<br></li>
    <li>34650<br></li>
    <li>39200<br></li> 
    <li>42000</li> </ol>


In [17]:
nn_search(1200)

10 NN for Customer 1200


Customer ID  Similarity Score
41725        41726          0.954603
41663        41664          0.926438
23960        23961          0.923867
16982        16983          0.923040
8755          8756          0.920661
4307          4308          0.920348
16155        16156          0.919752
1105          1106          0.917888
13602        13603          0.917460
9815          9816          0.915249

In [18]:
nn_search(3650)

10 NN for Customer 3650


Customer ID  Similarity Score
6188          6189          0.950560
28140        28141          0.937784
8066          8067          0.936680
6017          6018          0.936171
5615          5616          0.927949
7306          7307          0.923025
20261        20262          0.918398
33474        33475          0.916447
23346        23347          0.913418
37972        37973          0.913042

In [19]:
nn_search(10400)

10 NN for Customer 10400


Customer ID  Similarity Score
17597        17598          0.964383
20959        20960          0.925819
26166        26167          0.918353
8652          8653          0.915153
28055        28056          0.914227
10515        10516          0.913437
9096          9097          0.911474
15890        15891          0.909768
11309        11310          0.909424
23686        23687          0.897207

In [20]:
nn_search(14930)

10 NN for Customer 14930


Customer ID  Similarity Score
33853        33854          0.957549
33176        33177          0.952029
2705          2706          0.938542
8230          8231          0.933491
237            238          0.932029
31460        31461          0.931961
4046          4047          0.930308
36759        36760          0.929137
4788          4789          0.928226
28024        28025          0.928155

In [18]:
nn_search(22330)

10 NN for Customer 22330


Customer ID  Similarity Score
9839          9840          0.931193
17136        17137          0.915372
22788        22789          0.915206
686            687          0.910131
22249        22250          0.907943
13616        13617          0.900969
21325        21326          0.895619
10588        10589          0.887767
22614        22615          0.884087
22883        22884          0.879431

In [19]:
nn_search(25671)

10 NN for Customer 25671


Customer ID  Similarity Score
17206        17207          0.961472
38035        38036          0.935065
23513        23514          0.932468
22336        22337          0.930909
17776        17777          0.922078
27991        27992          0.913298
26598        26599          0.912098
28053        28054          0.912002
41848        41849          0.910125
40431        40432          0.909962

In [20]:
nn_search(29311)

10 NN for Customer 29311


Customer ID  Similarity Score
24619        24620          0.835948
23959        23960          0.833641
34476        34477          0.826136
34630        34631          0.823541
3145          3146          0.822697
13990        13991          0.822025
7171          7172          0.818546
6764          6765          0.818267
25207        25208          0.817017
36153        36154          0.816659

In [21]:
nn_search(34650)

10 NN for Customer 34650


Customer ID  Similarity Score
6901          6902          0.931291
27642        27643          0.927901
20566        20567          0.924065
7657          7658          0.923906
30714        30715          0.922335
33301        33302          0.921600
31571        31572          0.921122
32192        32193          0.920330
24951        24952          0.920001
36354        36355          0.918987

In [22]:
nn_search(39200)

10 NN for Customer 39200


Customer ID  Similarity Score
13170        13171          0.958222
29358        29359          0.929250
37140        37141          0.928357
13046        13047          0.922020
9743          9744          0.921712
29462        29463          0.921251
27082        27083          0.920309
27578        27579          0.917305
42808        42809          0.912519
29054        29055          0.912070

In [23]:
nn_search(42000)

10 NN for Customer 42000


Customer ID  Similarity Score
32470        32471          0.958657
28026        28027          0.950954
31800        31801          0.943429
41315        41316          0.942950
39517        39518          0.937329
22741        22742          0.935395
41143        41144          0.935216
42418        42419          0.935024
32492        32493          0.933916
39424        39425          0.933633

### 4) Customer rating prediction

**1)**  Calculate the similarities between the given customer and all other customers and compute his 10-nn (most similar) customers.

**IMPORTANT: In the similarity calculations for this step you need to exclude the customer rating attribute.**

* We will use the same function as before but we will remove the customer rating attribute. from the calculation.

In [32]:
def nn_search_random(cust_id):

 x = len(bank)

 age_dis_matrix=np.zeros(shape=(1,x))
 job_dis_matrix=np.zeros(shape=(1,x))
 marital_dis_matrix=np.zeros(shape=(1,x))
 education_dis_matrix=np.zeros(shape=(1,x))
 default_dis_matrix=np.zeros(shape=(1,x))
 balance_dis_matrix=np.zeros(shape=(1,x))
 housing_dis_matrix=np.zeros(shape=(1,x))
 loan_dis_matrix=np.zeros(shape=(1,x))
 products_dis_matrix=np.zeros(shape=(1,x))

 i = cust_id

 for j in range(x):
        age_dis_matrix[0,j] = dis_numeric(i-1,j,'Age')
        job_dis_matrix[0,j] = dis_categorical(i-1,j,'Job')
        marital_dis_matrix[0,j] = dis_categorical(i-1,j,'Marital')
        education_dis_matrix[0,j] = dis_categorical(i-1,j,'Education')
        default_dis_matrix[0,j] = dis_categorical(i-1,j,'Default')
        balance_dis_matrix[0,j] = dis_numeric(i-1,j,'Balance')
        housing_dis_matrix[0,j] = dis_categorical(i-1,j,'Housing')
        loan_dis_matrix[0,j] = dis_categorical(i-1,j,'Loan')
        products_dis_matrix[0,j] = dis_jaccard(i-1,j)
    

 comb = 1- ((1/9)* (age_dis_matrix + job_dis_matrix + marital_dis_matrix + education_dis_matrix
                                        + default_dis_matrix + balance_dis_matrix + housing_dis_matrix +
                                         loan_dis_matrix + products_dis_matrix))

# Creating a dataframe with the similarity score
 comb = pd.DataFrame(np.transpose(comb))

# Adding the customer id
 comb['Customer ID'] = np.arange(comb.shape[0])+1
 comb.rename(columns = {0:'Similarity Score'}, inplace = True)

# Excluding the id of the target customer
 comb = comb.drop(comb[comb['Customer ID'] == i].index)

#Sorting the resulted data frame using the similarity score
 comb = comb.sort_values(by=['Similarity Score'], ascending = False)
# Keeping only the 10 first records 
 comb = comb.head(10)
#Change the column order of the dataframe 
 comb = comb.reindex(columns=['Customer ID','Similarity Score'])

#Print and return the wanted result
 print ('10 NN for Customer', i)
 return (comb)

* We are choosing the random customer 1997.

In [31]:
nn_search_random(1997)

10 NN for Customer 1997


Customer ID  Similarity Score
6423          6424          0.950919
16968        16969          0.944140
27181        27182          0.941381
35142        35143          0.936833
4455          4456          0.933831
34084        34085          0.931063
24284        24285          0.929438
17410        17411          0.928637
7211          7212          0.928039
31351        31352          0.926095

**2)** Based only on the 10 most similar customers computed in the previous step, predict
the customer rating rank using:
* The average rating rank of the 10 most similar customers (rounded to the
nearest integer).
* The weighted average rating rank of the 10 most similar customers (rounded
to the nearest integer).

* We will select the first column from the above output minus 1 in order to extract the line numbers of the nearest neighbours.

In [33]:
top10 = nn_search_random(1997).iloc[:,0]-1
top10

10 NN for Customer 1997


6423      6423
16968    16968
27181    27181
35142    35142
4455      4455
34084    34084
24284    24284
17410    17410
7211      7211
31351    31351
Name: Customer ID, dtype: int32

* Using the line numbers of the 10 nearest neighbours we will create a new data frame with the rows we selected.

In [34]:
df_top10 = bank.iloc[top10]
df_top10

Age         Job  Marital  Education Default  Balance Housing Loan  \
6423   35.0  technician  married  secondary      no      194     yes   no   
16968  27.0  technician  married  secondary      no        0     yes   no   
27181  27.0  technician  married  secondary      no     2735     yes   no   
35142  32.0  technician  married  secondary      no       91     yes   no   
4455   32.0  technician  married  secondary      no     3067     yes   no   
34084  26.0  technician  married  secondary      no      518     yes   no   
24284  50.0  technician  married  secondary      no       31     yes   no   
17410  37.0  technician  married  secondary      no     1063     yes   no   
7211   38.0  technician  married  secondary      no      226     yes   no   
31351  33.0  technician  married  secondary      no     9305     yes   no   

      Rating      Products  
6423       3      10,15,17  
16968      2            17  
27181      3            10  
35142      3            10  
4455       3            17  
34084      3  3,8,10,14,17  
24284      3       2,10,17  
17410      3            10  
7211       3            10  
31351      3     4,9,10,17

* Converting the `Rating` column values to numeric in order to be able to calculate the average.

In [35]:
df_top10["Rating"] = pd.to_numeric(df_top10["Rating"])

<ipython-input-35-e1a9d0f21a34>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_top10["Rating"] = pd.to_numeric(df_top10["Rating"])


* We will now calculate the average rating rank of the 10 most similar customers (rounded to the nearest integer)

In [36]:
average = round(df_top10["Rating"].mean())
average

3

* We will now create a new data frame combining the `df_top10` data frame we created before with the `Similarity Score` column for the 10 most similar customers.

In [37]:
df_top10_sim = pd.concat([df_top10, nn_search_random(1997).iloc[:,1]], axis=1)
df_top10_sim

10 NN for Customer 1997


Age         Job  Marital  Education Default  Balance Housing Loan  \
6423   35.0  technician  married  secondary      no      194     yes   no   
16968  27.0  technician  married  secondary      no        0     yes   no   
27181  27.0  technician  married  secondary      no     2735     yes   no   
35142  32.0  technician  married  secondary      no       91     yes   no   
4455   32.0  technician  married  secondary      no     3067     yes   no   
34084  26.0  technician  married  secondary      no      518     yes   no   
24284  50.0  technician  married  secondary      no       31     yes   no   
17410  37.0  technician  married  secondary      no     1063     yes   no   
7211   38.0  technician  married  secondary      no      226     yes   no   
31351  33.0  technician  married  secondary      no     9305     yes   no   

       Rating      Products  Similarity Score  
6423        3      10,15,17          0.950919  
16968       2            17          0.944140  
27181       3            10          0.941381  
35142       3            10          0.936833  
4455        3            17          0.933831  
34084       3  3,8,10,14,17          0.931063  
24284       3       2,10,17          0.929438  
17410       3            10          0.928637  
7211        3            10          0.928039  
31351       3     4,9,10,17          0.926095

* This is the final output of the newly created data frame:

In [38]:
df_top10_sim

Age         Job  Marital  Education Default  Balance Housing Loan  \
6423   35.0  technician  married  secondary      no      194     yes   no   
16968  27.0  technician  married  secondary      no        0     yes   no   
27181  27.0  technician  married  secondary      no     2735     yes   no   
35142  32.0  technician  married  secondary      no       91     yes   no   
4455   32.0  technician  married  secondary      no     3067     yes   no   
34084  26.0  technician  married  secondary      no      518     yes   no   
24284  50.0  technician  married  secondary      no       31     yes   no   
17410  37.0  technician  married  secondary      no     1063     yes   no   
7211   38.0  technician  married  secondary      no      226     yes   no   
31351  33.0  technician  married  secondary      no     9305     yes   no   

       Rating      Products  Similarity Score  
6423        3      10,15,17          0.950919  
16968       2            17          0.944140  
27181       3            10          0.941381  
35142       3            10          0.936833  
4455        3            17          0.933831  
34084       3  3,8,10,14,17          0.931063  
24284       3       2,10,17          0.929438  
17410       3            10          0.928637  
7211        3            10          0.928039  
31351       3     4,9,10,17          0.926095

* We will calculate the weighted average rating rank of the 10 most similar customers (rounded to the nearest integer) by using columns `Rating` and `Similarity Score`.

In [39]:
def weighted_average(df_top10_sim, values, weights):
    return sum(df_top10_sim[weights] * df_top10_sim[values]) / df_top10_sim[weights].sum()
print(round(weighted_average(df_top10_sim, 'Rating', 'Similarity Score')))

3


**3)** For the evaluation of your classification algorithm you will use the 50 first records of the bank dataset and predict the rating for them. Then, for all n=50 records calculate the Mean Prediction Error for both prediction methods.

* We create a new data set with the first 50 records of the bank dataset.

In [40]:
evaluate = bank.head(50)
evaluate

Age            Job   Marital  Education Default  Balance Housing Loan  \
0   33.0   entrepreneur   married  secondary      no        2     yes  yes   
1   35.0     management   married   tertiary      no      231     yes   no   
2   41.0     management    single   tertiary      no      447     yes  yes   
3   42.0   entrepreneur  divorced   tertiary     yes        2     yes   no   
4   58.0        retired   married    primary      no      121     yes   no   
5   43.0     technician    single  secondary      no      593     yes   no   
6   41.0         admin.  divorced  secondary      no      270     yes   no   
7   29.0         admin.    single  secondary      no      390     yes   no   
8   53.0     technician   married  secondary      no        6     yes   no   
9   41.0       services   married  secondary      no      162     yes   no   
10  51.0        retired   married    primary      no      229     yes   no   
11  41.0    blue-collar   married    primary      no       52     yes   no   
12  60.0        retired   married    primary      no       60     yes   no   
13  33.0       services   married  secondary      no        0     yes   no   
14  28.0    blue-collar   married  secondary      no      723     yes  yes   
15  56.0     management   married   tertiary      no      779     yes   no   
16  32.0    blue-collar    single    primary      no       23     yes  yes   
17  41.0       services   married  secondary      no       50     yes   no   
18  40.0        retired   married    primary      no        0     yes  yes   
19  44.0         admin.   married  secondary      no     -372     yes   no   
20  39.0     management    single   tertiary      no      255     yes   no   
21  52.0   entrepreneur   married  secondary      no      113     yes  yes   
22  46.0     management    single  secondary      no     -246     yes   no   
23  36.0     technician    single  secondary      no      265     yes  yes   
24  57.0     technician   married  secondary      no      839      no  yes   
25  49.0     management   married   tertiary      no      378     yes   no   
26  60.0         admin.   married  secondary      no       39     yes  yes   
27  59.0    blue-collar   married  secondary      no        0     yes   no   
28  51.0     management   married   tertiary      no    10635     yes   no   
29  57.0     technician  divorced  secondary      no       63     yes   no   
30  25.0    blue-collar   married  secondary      no       -7     yes   no   
31  53.0     technician   married  secondary      no       -3      no   no   
32  36.0         admin.  divorced  secondary      no      506     yes   no   
33  37.0         admin.    single  secondary      no        0     yes   no   
34  44.0       services  divorced  secondary      no     2586     yes   no   
35  50.0     management   married  secondary      no       49     yes   no   
36  54.0        retired   married  secondary      no      529     yes   no   
37  36.0         admin.    single    primary      no     -171     yes   no   
38  58.0  self-employed   married   tertiary      no     -364     yes   no   
39  44.0     technician   married  secondary      no        0     yes   no   
40  55.0     technician  divorced  secondary      no        0      no   no   
41  29.0     management    single   tertiary      no        0     yes   no   
42  54.0    blue-collar   married  secondary      no     1291     yes   no   
43  48.0     management  divorced   tertiary      no     -244     yes   no   
44  32.0     management   married   tertiary      no        0     yes   no   
45  42.0         admin.    single  secondary      no      -76     yes   no   
46  24.0     technician    single  secondary      no     -103     yes  yes   
47  38.0   entrepreneur    single   tertiary      no      243      no  yes   
48  38.0     management    single   tertiary      no      424     yes   no   
49  46.0       services   married    primary      no      179     yes   no   

   Rating                 

* We need to convert the `Rating` column to numeric in order to be able to calculate the MPE.

In [41]:
evaluate["Rating"] = pd.to_numeric(evaluate["Rating"])

<ipython-input-41-cccb127c7a2c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  evaluate["Rating"] = pd.to_numeric(evaluate["Rating"])


* We perform some corrections to function `dis_numeric`, because with the new dataset the range changes.

In [42]:
def dis_numeric2(a,b,col):
    return abs((evaluate[col].iloc[a])-(evaluate[col].iloc[b]))/(int(max(evaluate[col]))-int(min(evaluate[col])))
#This function computes for a,b numeric then d(a,b) = |a-b| / maxvalue - minvalue
dis_numeric2(1,3,'Age')

0.19444444444444445

* We create the new function for the `evaluate` data set we created.

In [43]:
def nn_search_50(cust_id):

 x = len(evaluate)

 age_dis_matrix=np.zeros(shape=(1,x))
 job_dis_matrix=np.zeros(shape=(1,x))
 marital_dis_matrix=np.zeros(shape=(1,x))
 education_dis_matrix=np.zeros(shape=(1,x))
 default_dis_matrix=np.zeros(shape=(1,x))
 balance_dis_matrix=np.zeros(shape=(1,x))
 housing_dis_matrix=np.zeros(shape=(1,x))
 loan_dis_matrix=np.zeros(shape=(1,x))
 products_dis_matrix=np.zeros(shape=(1,x))

 i = cust_id

 for j in range(x): 
        age_dis_matrix[0,j] = dis_numeric2(i-1,j,'Age')
        job_dis_matrix[0,j] = dis_categorical(i-1,j,'Job')
        marital_dis_matrix[0,j] = dis_categorical(i-1,j,'Marital')
        education_dis_matrix[0,j] = dis_categorical(i-1,j,'Education')
        default_dis_matrix[0,j] = dis_categorical(i-1,j,'Default')
        balance_dis_matrix[0,j] = dis_numeric2(i-1,j,'Balance')
        housing_dis_matrix[0,j] = dis_categorical(i-1,j,'Housing')
        loan_dis_matrix[0,j] = dis_categorical(i-1,j,'Loan')
        products_dis_matrix[0,j] = dis_jaccard(i-1,j)
    

 comb = 1-((1/9)* (age_dis_matrix + job_dis_matrix + marital_dis_matrix + education_dis_matrix
                                        + default_dis_matrix + balance_dis_matrix + housing_dis_matrix +
                                         loan_dis_matrix + products_dis_matrix))

 comb = pd.DataFrame(np.transpose(comb))


 comb['Customer ID'] = np.arange(comb.shape[0])+1
 comb.rename(columns = {0:'Similarity Score'}, inplace = True)

 
 comb = comb.drop(comb[comb['Customer ID'] == i].index)

 comb = comb.sort_values(by=['Similarity Score'], ascending=False)
 comb = comb.head(10)
 comb = comb.reindex(columns=['Customer ID','Similarity Score'])

 return (comb)

In [44]:
nn_search_50(3)

Customer ID  Similarity Score
20           21          0.797445
48           49          0.768286
41           42          0.748574
1             2          0.673745
22           23          0.672017
23           24          0.663286
15           16          0.661463
47           48          0.655348
25           26          0.641279
43           44          0.638086

* Calculating the Mean Prediction Error for the two methods in order to evaluate them.

In [45]:
data = []
data2= []
for i in range(1,51):
    top10 = nn_search_50(i).iloc[:,0]-1
    df_top10 = evaluate.iloc[top10]
    average = round(df_top10['Rating'].mean())
    #print('The average rating for Customer',i, ':', average, evaluate['Rating'][i-1])
    MPE_average = average - evaluate['Rating'][i-1]
    #print(MPE)
    data.append(abs(MPE_average))
    df_top10_sim = pd.concat([df_top10, nn_search_50(i).iloc[:,1]], axis=1)
    weighted_average = round(sum(df_top10_sim['Similarity Score'] * df_top10_sim['Rating']) / df_top10_sim['Similarity Score'].sum())
    MPE_weight_average = weighted_average - evaluate['Rating'][i-1]
    data2.append(abs(MPE_weight_average))
print('The MPE for the average method is:',sum(data)/50)
print('The MPE for the weighted average method is:',sum(data2)/50)

The MPE for the average method is: 0.42
The MPE for the weighted average method is: 0.46


From the Mean Prediction Error we calculated for the two methods, we can clearly see that the best method is the average, as the MPE is 0.42 which is lower than the MPE of the weighted average method.